File: java/src/main/core/TokenBucket.java
package main.core;

import main.time.TimeProvider;

/*
 * TokenBucket implements token bucket rate limiting.
 * Each request or call consume a token and token are refilled after a fixed time interval
 */

public class TokenBucket {
    private final int bucketCapcity;
    private int tokens;
    private final long refillInterval;
    private long lastAccessTime;
    private long lastRefillTime;
    private final TimeProvider timeProvider;

    public TokenBucket(int bucketCapacity, long refillInterval, TimeProvider timeProvider) {
        this.bucketCapcity = bucketCapacity;
        this.tokens = bucketCapacity;
        this.lastRefillTime = timeProvider.now();
        this.timeProvider = timeProvider;
        this.refillInterval = refillInterval;
        this.lastAccessTime = lastRefillTime;
    };

    /*
     * Consume one token from the bucket.
     * 
     * @return true if token was consumed, false if limit exceeded
     */

    public synchronized boolean consumeToken() {
        lastAccessTime = timeProvider.now();

        refillBucketIfNeeded();

        if (tokens <= 0)
            return false;
        tokens--;
        return true;
    };

    /*
     * Refills the bucket if refill interval has elapsed
     */

    private void refillBucketIfNeeded() {
        long now = timeProvider.now();
        long elapsed = now - lastRefillTime;

        if (elapsed >= refillInterval) {
            tokens = bucketCapcity;
            lastRefillTime = now;
        }
    };

    /*
     * @return last access time in ms of token bucket
     */

    public long getLastAccessTime() {
        return lastAccessTime;
    }
}

File: typescript/src/core/TokenBucket.ts
import { TimeProvider } from "../time/TimeProvider";

/* TokenBucket implements token bucket rate limiting
 * Each request consume a token and tokens are refilled at fixed time interval
 */

export class TokenBucket {
  private readonly bucketCapacity: number;
  private tokens: number;
  private readonly refillInterval: number;
  private lastAccessTime: number;
  private lastRefillTime: number;
  private timeProvider: TimeProvider;

  constructor(
    bucketCapacity: number,
    refillInterval: number,
    timeProvider: TimeProvider
  ) {
    this.bucketCapacity = bucketCapacity;
    this.tokens = this.bucketCapacity;
    this.refillInterval = refillInterval;
    this.lastRefillTime = timeProvider.now();
    this.lastAccessTime = this.lastRefillTime;
    this.timeProvider = timeProvider;
  }

  /* @return true if token was consumed, false if limit exceeded */

  public consumToken(): boolean {
    this.lastAccessTime = this.timeProvider.now();

    this.refillBucketIfNeeded();

    if (this.tokens <= 0) {
      return false;
    }

    this.tokens--;
    return true;
  }

  /* refill the bucket if refill interval has elapsed */

  private refillBucketIfNeeded(): void {
    const currentTime = this.timeProvider.now();
    const elapsed = currentTime - this.lastRefillTime;

    if (elapsed >= this.refillInterval) {
      this.tokens = this.bucketCapacity;
      this.lastRefillTime = currentTime;
    }
  }

  /* @return last access time in ms of token bucket */

  public getLastAccessTime(): number {
    return this.lastAccessTime;
  }
}
